Transcript, with preliminary proofreading (#104)
fractureiser-investigation/fractureiser: docs/2023-06-08-meeting-transcript.md

# Transcript of 2023-06-08 meeting

## Opening 

[Start at 00:00:00]

Emi: Hello everyone, I'm Emi, and I'm going to be leading this meeting, representing the Fractureiser Mitigation Team. 

> This is gonna be a discussion with community memebers of note about the mitigation for potential future malware incidents.

Author: *mfw rolling keyboard without looking at keyboard*

As a summary, Fractureiser is a 4-Stage, self-updating, self-replicating piece of malware that boardly targets Java programs with the focus on Minecraft and modded communities. 

Stage3 is capable of stealing all browser information, including passwords, payment info, and cookies. It's also capable of stealing Microsoft, Minecraft and Discord credentials along with cryptocurrency. 

> Contained is also various botnet utilties for DDoS-ing, the ability to run arbirtrary system commands, and all stuff like that. It's a comprehnesive, uh, huh, on the entire system. 

Reviewer: ```suggestion
Contained is also various botnet utilities for DDoS-ing, the ability to run arbitrary system commands, and all stuff like that. It's a comprehensive pwn on the entire system. 
```

I kinda wrote up this agenda, that you, uh that is linked in the GitHub, and I guess, I've ordered these topics, 
there's about 4 or 5 of them, from the most realistic or actionable to the most kinda like, "pie in the sky". 

So if we run out of time, we'll probably cut some of the later ones. 

The first one I'd like to start with, I guess, just from the perspective of the community or the incident response, was the questions like what actually happens when a mod repository "reviews" the mod that is submitted. Are there new mod or an update? And what sort of checks are being run, what should be put in place, and those are sort of the question that I'd like to discuss in the first point. 

Back to you, Emi or Jasmine.

Emi: sure, it looks like Fury has requested to speak, so we'll give them the floor. 

Fury: Hi everybody. My name is Fury, and, CEO of Overwolf. Emi, before I touch on this uh first topic, um sorry willie. 

I just want to say, we're... I'm very grateful, we're all very grateful in Overwolf and CurseForge, for all the help we received from a lot of the folks here, 

> Super ??? in my opinion, that there is such involvement, dedication and commitment on behalf of the community. The way the things happened, I agree that if this is not caught on time, the impact could've been much more severe.  

Reviewer: I also have no idea what they're saying here.

Author: > I don't have the audio, but I'd say "nice", "cool" or "awesome" would fit

Quote the doc:

> You can watch the recording on [YouTube](https://www.youtube.com/watch?v=L52Hu334Q90) or [PeerTube](https://tube.sleeping.town/w/c48e7df1-cf9b-43d2-84a8-4bce404ee836).

Reviewer 2: sounds like "intriguing" to me

Reviewer 3: imo sounds like the word ends in -ling. Why don't we ping those people to decipher their own ~~writing~~ speech?

Um, in terms of the manual checks, whenever something is triggered, um from these automation checks, then a human would then review the mod. By the way, if someone changes the texts and images, we also require manual review, to avoid profanity and that content, all of these things, um.  

But if those automated checks have trigged some sort of the issue, or there is suspicous, it would be an manual check. 

Also I wanna emphasis that this is continuously evolving process for us, uh so just for example following this incident, we already updated our automatic checks, to um test for this particular vulnerabiliy, and should we identified this particular vulnerability again, it would trigger a manual review, um because we want to double check to make sure that, um, the mod are not compromised.    

I also saw the idea of um [DISCONNECTED]

Emi: ah, looks like they've muted. We also have the Geometrically from Modrinth who's here to speak, so uh, if you have any finishing words, go ahead. We want other voices. 

Fury: sorry, uh I got disconnected I think. Can you guys hear now?
 
Emi: yes.

Fury: OK. Can you repeat? I did not listen; I am not here for the first thing what's that?

> Emi: Since you muted, and because we want to get as many voices in here as possible, we're going to [be] moving on to modern's perspective, if you have any quick final words, feel free.

Reviewer: ```suggestion
Emi: Since you muted, and because we want to get as many voices in here as possible, we're going to [be] moving on to Modrinth's perspective, if you have any quick final words, feel free.
```

Author: *faceplam moment*

Fury: On this topic of which tests were running or in general?

Emi: Yes.

Fury: Um, let me see... Yeah I mean, there are couple more things here but you don't have to have me have microphones. That was too long, so no worries.

Emi: sounds good.

Geo: hey all, can you hear me?

Emi: Yes.

> Geo: Hey, I'm Jay, I'm the founder of Modrinth, and yeah, it's great to be here, it's really cool to kinda of see all sides of the community really come together. 

Author: Geo indeed used "Jay" here in the meeting. I believe the meeting agenda/minutes also list "Jay" alongside with "Geometrically".

So I am assuming some part of the modding toolchain has non-determinism that we can ruled out, and I think it is pretty actionable. Just examinating every part of the toolchain, making sure we don't violate those assumptions. 

And the crazy idea on the agenda, Geo kinda covered it already, that F-droid-style: we submit the source and it's get built by the mod repo. 

But, as we kinda covered this, this is kinda unrealistic, so I am not gonna spend too much time talking about that. 

So I mainly look for discussion about the stuff I mentioned before, which is making our default templates and the toolchain more, uh, easy to verify and reproduce. 

Emi: sounds great, we have a representative from the Quilt project here that would like to speak on this topic. And I will give them the floor. 

> Starchild: Hello, so very quickly, I am the admin and ??? lead for Quilt, also I am DevOp experts in my day jobs. 

Reviewer: ```suggestion
Starchild: Hello, so very quickly, I am the admin and infrastructure lead for Quilt, also I am DevOp experts in my day jobs. 
```

I think that's only two things that are really relevant, maybe someone else has other ideas, but I don't really see how we can work on something else that is actually productive and actually to do something.

Emi: thank you for your thoughts on that. Uh, we have a member of the Fractureiser team that would like to say something on code signing. 

Jasmine: Yeah. Hello, I'm Jasmine. 

So uh, my thought on code signing, is the one from where I recently uploaded something to the Maven Central repository, which enforces code signing. 

And that process was mostly able to be done through, like the build script itself; the most painful part of the process was getting the documentation on how to do it.  

So I feel like if we're able to get something done, where it's part of the Gradle system itself, then it might reduce the ... yeah, like sort of the friction between like, um, enable to do the code signing, and uh, like, uploading onto mod distribution sites. 

But that's something we are gonna have to work together with various of the modding projects to get it done for example mods and such. 

> Emi: I'd like to bring up some of these things that have been messaged in the meeting's channel, because there are people talking in text chant not speaking on voice, so that's not being recorded, and I'd like to relay it. 

Reviewer 2: ```suggestion
Emi: I'd like to bring up some of these things that have been messaged in the meeting's channel, because there are people talking in text chat not speaking on voice, so that's not being recorded, and I'd like to relay it. 
```

cpw: There are mechanisms. Uh, we are.. probably needed to publish guidelines on how you could do this stuff in parallel. But it would need commitment from... obviously, if you're committing to that, that's awesome, CurseForge is a very big distribution platform. I hope we can get similar commitment from Modrinth guys. Um, yeah.  

Emi: Alright, I'm definitely invite Modrinth to speak; they have thoughts on this topic.

Geo: Yeah. Yeah, I think like, personally, this is not related to Modrinth at all, I think it would be best for like a trusted third party to do it, mainly just because like the platform as a whole, we both have, theirs like, moderations rights. 

Let's say we banned a user because like, let's say they violated our content policy or uploading copyrighted content or something like that, right? 

We wouldn't want them to lose their ability to code-sign from a certificate authority. 

I don't know like how we could do that. I guess the fact that someone uploading to a user account, we have to associatie that. 

> But I would agree definitely that we'd love to work on code-sigining, and kinda of becoming a certificate authority. 

Reviewer 2: ```suggestion
But I would agree definitely that we'd love to work on code-sigining, and kind of becoming a certificate authority. 
```

The agenda has some light discussion for macOS and Linux, and people are kinda building prototypes on that, which is cool, we can do that where it's possible, but I'd urge people have to focus on Winodws cause that's where all the players are.  

Just the state of things there, doesn't look that great. 

So, I don't know if there is actually that much to discuss, but since it was brought up multiple times during the whole incident, I thought to bring it up here again. 

One of the other thing, that I'm just gonna mention out of the gate, is that JVM-level sandboxing is just, not gonna work. Besides the fact that it is getting deprecated and removed in a few years, it's just broken, so don't suggest that. 

Alright, so open the floor for discussion. 

Emi: Yeah, we have a couple of people from the Prism Launcher team that would like to speak on this topic, so I am gonna invite them in to have their perspective on this issue. 

ZekeZ: Good. Hi. How's everyone going today?

timoreo: Hello hello.

ZekeZ: Alright, um. Do you want to start this timoreo? You're the main, one of the main developers behind the actual launcher. 

> timoreo: Yeah. So I'll to start. So, sandboxing, first off, the first point: executing untruest will never be safe, no matter how much sandboxing you put, that's just a fact. [ZekeZ: um-hum.] You can never fully protect. 

Reviewer 2: ```suggestion
timoreo: Yeah. So I'll to start. So, sandboxing, first off, the first point: executing untrusted will never be safe, no matter how much sandboxing you put, that's just a fact. [ZekeZ: um-hum.] You can never fully protect. 
```

This is kind of, sandboxing is the last resort, just to limit the impact. 

In fact, we can do a bits of sandboxing because we have a distribution of Flatpak, which is sandboxed. 

It may seem a bit like a niche part, except that what's important is that Steam Deck users, which has actually a huge majority of their users, is forced to use Flatpak.

What Flatpak does is essentially restrict a lot of access, mainly file systems.  

So in this case, for this specific malware Fractureizer, it wouldn't have been able to steal information from the browser and all of that, or set up starting some malicious code that booted all of that.

ZekeZ: And yeah, in this case, we found it was ineffective, the malware, I believe, against the Flatpak. Is that correct?

timoreo: Yes. I mean, generally, the malware wasn't that well coded for Linux.

ZekeZ: This specific one, at least.

timoreo: This specific one was entirely broken on Linux. 

> But the problem is kind of, you can restrict more with Flatpaks, but you restrict ???. 

Reviewer: I think this is just them tripping over their words, and the next line is what they intended to say.

Author: Now I would assume the same. Fixed by replacing with "but... for example, you could restrict microphone, ...".

So I've heard some people throwing that about restricting what it connects to. That could be an interesting thought, except you can kind of connect to any server you want from multiplayer. 

So you would have to put some component inside of Minecraft, and as we already said, putting anything security inside of Minecraft is going to be pretty much... pretty much impossible. 

Yes, that's just one type of malware. It could still damage files, but with the files system restrictions, it could... block. But yeah.

Emi: All right. Thanks a lot for the team on these thoughts.

This is going to be a last call on thoughts on this topic. 

Once we hit the hour mark for our meeting, we're going to open the floor to all attendees and all agenda items to revisit any topics or talk about any aggregate items. 

ZekeZ: Thank you for having us on.

Emi: It's great to have you.

timoreo: Yeah, thank you.

### Overall recap

[Start at 00:59:29]

Emi: So if no one would like to speak on this, then yeah, we can move forward.  

We have a request from PaperMC, and I will given them floor.

> kennytv: Hello, kinda of our perspective from Paper. 

Reviewer 2: ```suggestion
kennytv: Hello, kind of our perspective from Paper. 
```

Emi: As a technical perspective from the Fractureiser Mitigation Team, it is our understanding that the Luna Pixel projects, the author that were compromised, did have Two-Factor Authentication enabled, and the reason this was able to be circumvented was session tokens, stored in the browser's history, which was sent to the malicious actor.  

But also yes, thank you for discussion on that, and I hope we see some form of 2FA on Modrinth in the future. 

Willie: yeah, I'm just gonna, again, kinda revisited some of the points that we kinda touched on here, one that I didn't really mention in the code-signing section, was that we should kinda be encouraging people to build, um, build and upload from CI. 

This was mentioned just now, but I didn't really get to that before. I think like that kinda mitigates the whole steal stuff from author's, from modder's local machine, because the signing token is not there, or the upload tokens or the signing keys aren't on the machine.  

And for GitHub Actions, they have this credential vault, where they also hide the private tokens to your GitHub Actions job, only for the trusted actions, and not PR that kind of stuff. 

> It's like very, it's pretty robust; people on the industry use it kinda of a bit.  

Reviewer 2: ```suggestion
It's like very, it's pretty robust; people on the industry use it kind of a bit.  
```

I feel... just stuff like that, get modders, teach modders about the build system more. 

That said, really badly, but I think that [Willie: Do you think...] could help.

Willie: Yeah do you think for beginners, it would probably be easy here to recommend GitHub actions right? 

Cause for Jenkins you need to setup the webhook like, have a separate system, have to talk to you, or someone in that hosts the Jenkins server. 

Um I think for beginners it's like, GitHub Actions is the easiest, safe by default grant. Once we get everything added to template, for stuff. 

Jared: For CI, yes. For the CD part, the Maven part, I am not too up-to-date on the GitHub whole thing, but I think for GitHub Packages you need to include your ... some credentials to be able just pull from it. 

So for the Jenkins part, yes. 

For everything, for the maven hosting part, not too sure on [it]. 

I know for me, I don't have a system set up in place to accept file uploads. 

All the binary on my maven have been built on the same machine, and I don't have a system to not do that, which is something I'm gonna look into after this whole system. Yes, good.

> Emi: Sure, I can definitely speak on that because I'm a author that uses GitHub Actions, and I use a thrid-party maven. 

Reviewer 2: ```suggestion
Emi: Sure, I can definitely speak on that because I'm a author that uses GitHub Actions, and I use a third-party maven. 
```

Typically, the situation from the GitHub CI perspective, is there is small amount of code built into build script, and the secret are passed from GitHub Actions, do you know, the username and password, as well as the maven URL, passed from GitHub Actions, and Gradle can just handle uploading files. 

I think in terms of accessibility, GitHub Actions publishing to maven is probably the most viable, and it's also worth considering that, for a lot of mods, maven is not strictly necessary. 

Um, there's just solutions like CurseMaven and Modrinth has a maven for all uploaded mods, which is useful for mods that aren't exposing themselves as APIs, and you know, for just like, casual add-on developers. 

> And... I think the space for API developers kinda of set off a higher standard for understanding the ecosystem, settup up Maven, and the communication there. 

Reviewer 2: ```suggestion
And... I think the space for API developers kina of set off a higher standard for understanding the ecosystem, setup up Maven, and the communication there. 
```
